src/auth/policy/resolver.go

```diff
@@ -1,41 +1,349 @@
 package policy
 
 import (
+	"fmt"
+	"strings"
+
+	"github.com/armon/go-radix"
 	"github.com/consensys/quorum-key-manager/src/auth/types"
 )
 
+var (
+	wildCardSuffix = "*"
+)
+
 type Operation struct {
 	Action string
 	Path   string
 }
 
-type Result struct {
-	isAllowed bool
-	err       error
+type Result interface {
+	Allowed() bool
+	Error() error
+}
+
+// Resolver is responsible to control whether an operation is authorized or not
+// depending on the set of policies attached to the resolver
+type Resolver interface {
+	IsAuthorized(op *Operation) Result
 }
 
-func (res *Result) Allowed() bool {
-	return res.isAllowed
+type statement struct {
+	types.Statement
+	policy *types.Policy
 }
 
-func (res *Result) Error() error {
-	return res.err
+type result struct {
+	op *Operation
+
+	allow *statement
+	deny  *statement
 }
 
-// Resolver is responsible to control whether an operation is authorized or not
-// depending on the set of policies attached to the resolver
-type Resolver struct {
-	// TODO: to be implemented
+func (res *result) Allowed() bool {
+	return res.deny == nil && res.allow != nil
+}
+
+func (res *result) Error() error {
+	if res.deny != nil {
+		return fmt.Errorf("action %q on resource %q denied by policy %q statement %q", res.op.Action, res.op.Path, res.deny.policy.Name, res.deny.Name)
+	}
+
+	if res.allow == nil {
+		return fmt.Errorf("action %q on resource %q not allowed", res.op.Action, res.op.Path)
+	}
+
+	return nil
+}
+
+// actionsResolver is associated to every statement path
+// it aggregates all statement actions associated to the path into a radix tree
+type actionsResolver struct {
```

Comment:
Why couldn't you use `radix.Tree` to allocate also your `exactAction` all your effectResolver? As I can quickly see from go-radix each node can store a value where we would store the effectResolver, is that right?

Reply:
In the radix tree all keys are prefixes. If we also include the exact path into tree then `Sign` would also give access to `SignTransaction` and we don't want that.

Reply:
Then it looks like Radix might not be a good fit since I would expect nodes to be split by a separator char, such as `/` or `:`. I would rather implement my own tree solution if that would be the case

As you describe you would have same issue with resources or action nesting other such as:
```
Action:
- Keys:List
- Keys:ListDeleted

Resources:
- stores::stores/mystoreOne/
- stores::stores/mystore
```

Reply:
Not sure to understand.

Radix is used for prefixes (with wildcard last character  `*`) in the case above those are exact path.

Reply:
But that is exactly what I am trying to avoid, using two independent flows for same feature. To mitigate this issue you describe above ` Sign would be treated as a prefix and would also give access to SignTransaction`, couldn't you just store as part of the node data (`effectResolver`) whether or not the path is a exactPath or a wildcarded one? so at the time you resolve you can deal with the issue of `Sign` vs `SignTransaction`